lib/shopify-cli/api.rb

```diff
@@ -0,0 +1,87 @@
+require 'shopify_cli'
+require 'net/http'
+
+module ShopifyCli
+  class API
+    include Helpers::OS
+    include SmartProperties
+
+    property! :ctx, accepts: ShopifyCli::Context
+    property! :token
+    property! :url
+
+    class APIRequestError < StandardError; end
+    class APIRequestNotFoundError < APIRequestError; end
+    class APIRequestClientError < APIRequestError; end
+    class APIRequestUnauthorizedError < APIRequestClientError; end
+    class APIRequestUnexpectedError < APIRequestError; end
+    class APIRequestRetriableError < APIRequestError; end
+    class APIRequestServerError < APIRequestRetriableError; end
+    class APIRequestThrottledError < APIRequestRetriableError; end
+
+    def self.gid_to_id(gid)
+      gid.split('/').last
+    end
+
+    def query(body, variables: {})
+      _, resp = request(
+        body,
+        variables: variables,
+        headers: default_headers,
+        graphql_url: url,
+      )
+      ctx.debug(resp)
+      resp
+    end
+
+    private
+
+    def request(body, graphql_url:, variables: {}, headers: {})
+      CLI::Kit::Util.begin do
+        uri = URI.parse(graphql_url)
+        http = ::Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = true
+
+        req = ::Net::HTTP::Post.new(uri.request_uri)
+        req.body = JSON.dump(query: body.tr("\n", ""), variables: variables)
+        req['Content-Type'] = 'application/json'
+        headers.each { |header, value| req[header] = value }
+        response = http.request(req)
+
+        case response.code.to_i
+        when 200..399
+          [response.code.to_i, JSON.parse(response.body)]
+        when 401
+          raise APIRequestUnauthorizedError, "#{response.code}\n#{response.body}"
+        when 404
+          raise APIRequestNotFoundError, "#{response.code}\n#{response.body}"
+        when 429
+          raise APIRequestThrottledError, "#{response.code}\n#{response.body}"
+        when 400..499
+          raise APIRequestClientError, "#{response.code}\n#{response.body}"
+        when 500..599
+          raise APIRequestServerError, "#{response.code}\n#{response.body}"
+        else
+          raise APIRequestUnexpectedError, "#{response.code}\n#{response.body}"
+        end
+      end.retry_after(APIRequestRetriableError, retries: 3) do |e|
+        sleep(1) if e.is_a?(APIRequestThrottledError)
+      end
+    end
+
+    def current_sha
+      output, status = ctx.capture2e('git', 'rev-parse', 'HEAD', chdir: ShopifyCli::ROOT)
+      status.success? ? output.strip : 'SHA unavailable'
+    end
+
+    def default_headers
+      {
+        'User-Agent' => "Shopify App CLI #{ShopifyCli::VERSION} #{current_sha} | #{uname(flag: 'v')}",
+      }.merge(auth_headers(token))
+    end
+
+    def auth_headers(token)
+      { Auth: token } # a nonsense header that is only meant to be replaced
```

Comment:
should we `raise NotImplementedError` here instead?

Reply:
Okay so I updated it so that it will raise NotImplementedError on if a named auth_header is not supplied and it isn't overridden. I did it this way because if I just raised, it would make the API class unusable unless it is inherited from.